### ISSUE TITLE
Service wont start on boot

### ISSUE THREAD

**Issue by rescaux** [aws/ec2-macos-init#9]
Dear ec2-macos-init contributors,

I've now tried multiple times (directly through launchd and through ec2-macos-init) to run a simple binary on system boot. This aims to join the machine to an Active Directory, etc... What I managed with this config so far was, that the script got executed correctly - but only a few seconds after a (random) user logged in to the machine - not on boot.

The config looks like this:
```
 [[Module]]
  Name = "SetupStartup"
  PriorityGroup = 4
  [Module.Command]
    Cmd = ["/usr/local/bin/startup.sh.x"]
    # I've tried this both with root and ec2-user
    RunAsUser = "ec2-user" 
```
chmod on the binary is a+x

**Comment by mattcataws** [comment]
Hey @rescaux, I'm sorry to hear that you're having issues with ec2-macos-init.

Would you be able to include any error message that was logged by ec2-macos-init? The log file's location is `/var/log/amazon/ec2/ec2-macos-init.log`.

Also, that's strange that the script was only executed after a user logged in... I've tried reproducing the problem with the config you posted and my own test script. However, I wasn't able to see the exact behavior that you're describing.

There was an error with the module you included -- it's missing a run type configuration (`RunOnce`, `RunPerBoot`, or `RunPerInstance`). Once I added that, ec2-macos-init was able to run my test script just fine. I'll include my test environment below.

My config looked like this:
```
[[Module]]
    Name = "SetupStartup"
    PriorityGroup = 4
    RunPerBoot = true
    [Module.Command]
        Cmd = ["/usr/local/bin/startup.sh.x"]
        RunAsUser = "ec2-user"
```

The contents of `/usr/local/bin/startup.sh.x`:
```
#!/bin/bash

echo "Hello World!"
```

The permissions I have on the script:
```
-rwxr-xr-x  1 ec2-user  admin  33 Apr 20 19:00 /usr/local/bin/startup.sh.x
```

The logs from priority level 4 after restarting the instance:
```
2022/04/20 19:15:26.399460 Processing priority level 4 (2 modules)...
2022/04/20 19:15:26.399472 Running module [SetupStartup] (type: command, group: 4)
2022/04/20 19:15:26.399489 Skipping module [ExecuteUserData] (type: userdata, group: 4) due to Run type setting
2022/04/20 19:15:26.451198 Successfully completed module [SetupStartup] (type: command, group: 4) with message: successfully ran command [[/usr/local/bin/startup.sh.x]] with stdout [Hello World!] and stderr []
2022/04/20 19:15:26.451252 Successfully completed processing of priority level 4
```

Could you try adding the run type configuration to your module and let me know what happens. Any error logs from that run would also be useful.

**Comment by rescaux** [comment]
Okay, I've now tried everything so far and it's still behaving weird. As one can see in [this ](https://youtu.be/CSm-JpysvSk)video, the script runs only when macos-ec2-init is invoked manually. It is invoked on boot as well, but won't do anything.

**Comment by jahkeup** [comment]
Hi @rescaux, thanks for sharing the logs~

> What I managed with this config so far was, that the script got executed correctly - but only a few seconds after a (random) user logged in to the machine - not on boot.

@mattcataws and I are wondering where in the script that this hangs. For example, does the script hang on the Domain Join, or when setting up the VPN connection? 

From the description provided and logs in your video seems, it like the script might be invoking a command that requires an active user login to complete (maybe a prompt, or initialization of some state by the OS). 

* Have you identified any particular command in this script that causes the script to hang?

  The logs that @mattcataws shared above have corresponding start/stop lines that the impacted instances' logs (briefly shown in your video) does not contain. These lines look like:

  ```
  2022/04/20 19:15:26.399472 Running module [SetupStartup] (type: command, group: 4)
          ... other Modules' messages ...
  2022/04/20 19:15:26.451198 Successfully completed module [SetupStartup] (type: command, group: 4) with message: successfully ran command [[/usr/local/bin/startup.sh.x]] with stdout [Hello World!] and stderr []
  ```

  If the script was able to complete its run, even if non-zero, we should see some corresponding log messages in `ec2-macos-init`'s logs.

* When you say the script runs "a few seconds after a (random) user logged in to the machine", what do you mean? Are these users SSH-ing into the host or are they VNC/NX-ing in?

  The act of (and method of) logging in might unblock the script, so this might help inform us of where to look next.

**Comment by rj93** [comment]
Hi @jahkeup, I also have a similar issue. I am using mac2.metal and ec2-macos-init version `1.5.3 [2022-06-09 16:07:51 -0700]` and in my ec2-macos-init I have the following:

```toml
### Group 6
## Custom services

[[Module]]
  Name = "StartGitlabRunner"
  PriorityGroup = 6
  RunPerBoot = true
  [Module.Command]
    Cmd = ["/Users/ec2-user/.gitlab-runner/run.sh"]
    RunAsUser = "ec2-user"
```

with my script as follows:

```sh
#!/bin/bash

if pgrep -x gitlab-runner >/dev/null
then
  echo "Skipping starting gitlab-runner as process is already running"
else
  echo "Starting gitlab-runner process..."
  /opt/homebrew/bin/gitlab-runner run &> /Users/ec2-user/.gitlab-runner/runner.log &
  echo "Started gitlab-runner process"
fi

echo "process id: $(pgrep -x gitlab-runner)"
```


The logs from the system restart are:

```
2022/07/15 13:41:37.774763 Fetching instance ID from IMDS...
2022/07/15 13:41:37.811484 Unable to get instance ID - IMDS may not be available yet...retrying every 1s [0/600]
2022/07/15 13:41:41.822631 Running on instance i-03b4cf7416cbc2031
2022/07/15 13:41:41.822802 Reading init config...
2022/07/15 13:41:41.829855 Successfully read init config
2022/07/15 13:41:41.829921 Validating config...
2022/07/15 13:41:41.830192 Successfully validated config
2022/07/15 13:41:41.830222 Prioritizing modules...
2022/07/15 13:41:41.830257 Successfully prioritized modules
2022/07/15 13:41:41.830281 Creating instance history directories for current instance...
2022/07/15 13:41:41.830420 Successfully created directories
2022/07/15 13:41:41.830449 Getting instance history...
2022/07/15 13:41:41.838421 Successfully gathered instance history
2022/07/15 13:41:41.838481 Processing priority level 1 (2 modules)...
2022/07/15 13:41:41.839368 Running module [UnmountLocalSSD] (type: command, group: 1)
2022/07/15 13:41:41.844034 Running module [DisableEthernet] (type: command, group: 1)
2022/07/15 13:41:43.131293 Successfully completed module [UnmountLocalSSD] (type: command, group: 1) with message: successfully ran command [[/bin/zsh -c diskutil list internal physical | egrep -o '^/dev/disk\d+' | xargs diskutil eject || true]] with stdout [] and stderr [Volume failed to eject]
2022/07/15 13:41:53.761170 Successfully completed module [DisableEthernet] (type: command, group: 1) with message: successfully ran command [[/usr/sbin/networksetup -setnetworkserviceenabled Ethernet off]] with stdout [] and stderr []
2022/07/15 13:41:53.761301 Successfully completed processing of priority level 1
2022/07/15 13:41:53.761326 Processing priority level 2 (1 modules)...
2022/07/15 13:41:53.761374 Running module [CheckNetworkIsUp] (type: networkcheck, group: 2)
2022/07/15 13:41:53.772429 Successfully completed module [CheckNetworkIsUp] (type: networkcheck, group: 2) with message: successfully pinged default gateway with a RTT of 368µs
2022/07/15 13:41:53.772520 Successfully completed processing of priority level 2
2022/07/15 13:41:53.772545 Processing priority level 3 (11 modules)...
2022/07/15 13:41:53.772609 Running module [DisableWiFi] (type: command, group: 3)
2022/07/15 13:41:53.772628 Running module [EC2SuggestedDefaultConfigPerformance] (type: systemconfig, group: 3)
2022/07/15 13:41:53.772680 Skipping module [ManageEC2User] (type: usermanagement, group: 3) due to Run type setting
2022/07/15 13:41:53.772720 Running module [EC2SuggestedDefaultConfigSecurity] (type: systemconfig, group: 3)
2022/07/15 13:41:53.772810 Skipping module [SetDefaultTimezone] (type: command, group: 3) due to Run type setting
2022/07/15 13:41:53.773784 Skipping module [DisableSleep] (type: command, group: 3) due to Run type setting
2022/07/15 13:41:53.773854 Skipping module [NeverSleep] (type: command, group: 3) due to Run type setting
2022/07/15 13:41:53.773885 Skipping module [NeverSleepDisplay] (type: command, group: 3) due to Run type setting
2022/07/15 13:41:53.774044 Running module [UpdateMOTD] (type: motd, group: 3)
2022/07/15 13:41:53.774129 Skipping module [RemoveSSHGroup] (type: command, group: 3) due to Run type setting
2022/07/15 13:41:53.772618 Skipping module [SetAmazonTimeSync] (type: command, group: 3) due to Run type setting
2022/07/15 13:41:53.774703 Did not modify SSHD configuration
2022/07/15 13:41:53.787156 Successfully completed module [EC2SuggestedDefaultConfigSecurity] (type: systemconfig, group: 3) with message: system configuration completed with [0 changed / 1 unchanged / 0 error(s)] out of 1 requested changes
2022/07/15 13:41:53.800800 Modified sysctl property [net.inet.tcp.autosndbufmax=33554432]
2022/07/15 13:41:53.802630 Modified sysctl property [net.inet.tcp.sendspace=1048576]
2022/07/15 13:41:53.802706 Modified sysctl property [net.link.generic.system.rcvq_maxlen=1024]
2022/07/15 13:41:53.803838 Modified sysctl property [kern.aiothreads=64]
2022/07/15 13:41:53.803945 Modified sysctl property [net.inet.tcp.win_scale_factor=8]
2022/07/15 13:41:53.804990 Did not modify default [AutomaticCheckEnabled]
2022/07/15 13:41:53.805009 Did not modify default [AutomaticDownload]
2022/07/15 13:41:53.805111 Did not modify default [AutomaticallyInstallMacOSUpdates]
2022/07/15 13:41:53.805242 Modified sysctl property [net.inet.tcp.autorcvbufmax=33554432]
2022/07/15 13:41:53.805271 Modified sysctl property [kern.aiomax=900]
2022/07/15 13:41:53.805493 Did not modify default [CriticalUpdateInstall]
2022/07/15 13:41:53.805659 Did not modify default [ConfigDataInstall]
2022/07/15 13:41:53.806199 Modified sysctl property [net.inet.tcp.recvspace=1048576]
2022/07/15 13:41:53.814463 Successfully completed module [UpdateMOTD] (type: motd, group: 3) with message: successfully updated motd file [/etc/motd] with version string [macOS Monterey 12.4]
2022/07/15 13:41:53.837840 Error while running module [DisableWiFi] (type: command, group: 3) with message:  and err: ec2macosinit: error executing command [[/bin/zsh -c wifidevice="$(networksetup -listallhardwareports | grep -A 1 "Wi-Fi" | tail -n 1 | cut -d " " -f2)"; if [[ ! -z $wifidevice ]]; then networksetup -setairportpower $wifidevice off; fi]] with stdout [Wireless is not available.
** Error: Wireless is not currently available on this machine.] and stderr []: exit status 9
2022/07/15 13:41:53.905767 Modified sysctl property [kern.aioprocmax=256]
2022/07/15 13:41:53.905875 Successfully completed module [EC2SuggestedDefaultConfigPerformance] (type: systemconfig, group: 3) with message: system configuration completed with [9 changed / 5 unchanged / 0 error(s)] out of 14 requested changes
2022/07/15 13:41:53.905917 Successfully completed processing of priority level 3
2022/07/15 13:41:53.905935 Processing priority level 4 (1 modules)...
2022/07/15 13:41:53.905968 Skipping module [GetSSHKeys] (type: sshkeys, group: 4) due to Run type setting
2022/07/15 13:41:53.905986 Successfully completed processing of priority level 4
2022/07/15 13:41:53.906000 Processing priority level 5 (1 modules)...
2022/07/15 13:41:53.906021 Skipping module [ExecuteUserData] (type: userdata, group: 5) due to Run type setting
2022/07/15 13:41:53.906036 Successfully completed processing of priority level 5
2022/07/15 13:41:53.906046 Processing priority level 6 (1 modules)...
2022/07/15 13:41:53.906067 Running module [StartGitlabRunner] (type: command, group: 6)
2022/07/15 13:41:53.963004 Successfully completed module [StartGitlabRunner] (type: command, group: 6) with message: successfully ran command [[/Users/ec2-user/.gitlab-runner/run.sh]] with stdout [Starting gitlab-runner process...
Started gitlab-runner process
process id: 433] and stderr []
2022/07/15 13:41:53.963070 Successfully completed processing of priority level 6
2022/07/15 13:41:53.963082 Writing instance history for instance i-03b4cf7416cbc2031...
2022/07/15 13:41:53.969077 Successfully wrote instance history
2022/07/15 13:41:53.969110 EC2 macOS Init completed in 12.151719875s
```

Which suggests that the server is actually starting, but then exits.

I've tried adding a `sleep 600` to the end of the `run.sh` file and when SSHing onto the instance I can see the process running:

```
ec2-user@ip-10-211-190-39 ~ % ps -ef | grep gitlab
  501   428   109   0  1:55pm ??         0:00.00 /bin/bash /Users/ec2-user/.gitlab-runner/run.sh
  501   431   428   0  1:55pm ??         0:00.31 /opt/homebrew/bin/gitlab-runner run
  501   503   487   0  1:56pm ttys000    0:00.00 grep gitlab
```

Waiting for the `sleep 600` to finish and the the process is killed:

```
ec2-user@ip-10-211-190-39 ~ % ps -ef | grep gitlab
  501   656   487   0  2:06pm ttys000    0:00.00 grep gitlab
 ```
 
 logs:
 ```
 2022/07/15 13:55:41.302136 Running module [StartGitlabRunner] (type: command, group: 6)
2022/07/15 14:05:41.601508 Successfully completed module [StartGitlabRunner] (type: command, group: 6) with message: successfully ran command [[/Users/ec2-user/.gitlab-runner/run.sh]] with stdout [Starting gitlab-runner process...
Started gitlab-runner process
process id: 431] and stderr []
2022/07/15 14:05:41.601820 Successfully completed processing of priority level 6
2022/07/15 14:05:41.601866 Writing instance history for instance i-03b4cf7416cbc2031...
2022/07/15 14:05:41.608182 Successfully wrote instance history
2022/07/15 14:05:41.608241 EC2 macOS Init completed in 10m14.065216375s
```

**Comment by jahkeup** [comment]
Thanks for reporting in and posting logs @rj93!

We suspect you're up against a separate (but related) issue that comes into play when processes spawn in the context of a Launchd Daemon (and/or Agents). [Processes created by Launchd Daemons (directly or indirectly) must not make system calls that would spawn a whole other _conceptual_ "daemon process" according to the "Required Behaviors" section of Launchd documentation](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html#//apple_ref/doc/uid/10000172i-SW7-104302). It is likely you'll need a separate Launchd Daemon instance that launches and owns the GitLab CI runner process (set up with your flags and environment as needed).

As written, the shared script creates a "background" process (from the `gitlab-runner run ... &` command), that belongs to the same session as `ec2-macos-init` and terminates/exits along with the primary Launchd process - ie: `ec2-macos-init`.

@okudajun and I tried alternatives to spawn from executed scripts and still found that they failed to stick around or failed to run at all (the usual trustworthy `nohup` had its syscalls fail as inappropriate IOCTLs). My knee-jerk suggestion would have been to try with `nohup`... but.. doesn't appear to work (as expected based on [docs](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html)).

Our team will need to look at how `ec2-macos-init` itself spawns and what it executes from new angles to determine if `ec2-macos-init` can transparently support or have helpers to support managing ad-hoc daemon processes. As written today, however, and based on the [Launchd - Required Behavior documentation](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html#//apple_ref/doc/uid/10000172i-SW7-104142-BCICCCFI), `ec2-macos-init` cannot spawn daemon processes.

---

@rescaux - is there a process spawned in `startup.sh` that might want to daemonize? If so, you might need to spawn a Launchd Daemon (or multiple, dependent on your use case) in order to work around this Launchd limitation.

**Comment by rj93** [comment]
Hi @jahkeup, it may be worth me pointing out that this `run.sh` script is also called from userdata at instance launch and it keeps the process running once the `ec2-macos-init` process has completed.

Heres a snippet from my userdata:

```bash

# install gitlab-runner, etc
# ...

echo "Starting gitlab-runner..."
su - ec2-user -c '/Users/ec2-user/.gitlab-runner/run.sh'
echo "Started gitlab-runner"

# Won't run run as part of this ec2-macos-init process, but if the instance reboots will start the gitlab-runner
EC2_MACOS_INIT_FILE=/usr/local/aws/ec2-macos-init/init.toml
if ! grep -q "StartGitlabRunner" $EC2_MACOS_INIT_FILE
then
  echo "Adding group ec2-macos-init..."
  cat >> $EC2_MACOS_INIT_FILE <<-EOF

	### Group 6
	## Custom services

	[[Module]]
	  Name = "StartGitlabRunner"
	  PriorityGroup = 6
	  RunPerBoot = true
	  [Module.Command]
	    Cmd = ["/Users/ec2-user/.gitlab-runner/run.sh"]
	    RunAsUser = "ec2-user"
	EOF
  echo "Added group ec2-macos-init"
else 
  echo "Skipping adding group to ec2-macos-init"
fi

echo "Finished userdata script"

```

I was hoping to get away without having to use launchd, but thanks for the feedback I will go down that route instead.

**Comment by sourcedelica** [comment]
Hi @rj93 - what did you end up doing?  I'm trying to do the same thing (setup gitlab-runner to run at boot on a m2.metal instance) and ran into a different set of issues (trying to use `gitlab-runner install` which creates a launchd user agent which fails to start because there is no user session).  

>  It is likely you'll need a separate Launchd Daemon instance that launches and owns the GitLab CI runner process (set up with your flags and environment as needed).

 I'm confused about what is being recommended.  I created a launch daemon in `/Library/LaunchDaemons` that was basically the same contents of what `gitlab-runner install` created, plus `UserName ec2-user`.  This seems to work.  I'm assuming that the only issue with daemonization is running gitlab-runner via ec2-macos-init.  That running it as a regular launchd daemon is fine ([lack of Gitlab support](https://docs.gitlab.com/runner/install/osx.html#limitations-on-macos) notwithstanding).  Does that sound right?

Thanks!

**Comment by jahkeup** [comment]
> I created a launch daemon in /Library/LaunchDaemons that was basically the same contents of what gitlab-runner install created, plus UserName ec2-user.

Ah, sorry! You're right, yes, we are recommending to create a job - using user data - and then loading it to run under launchd. And yeah, that lack of support seems to me because of the particulars of creating and managing the LaunchDaemon and/or LaunchAgent. (to @sourcedelica , skip to the bottom of this reply then come back here!)

For general needs, it could be simple enough to run a LaunchDaemon in the `system` domain but then your `gitlab-runner` could be missing out on some services.. Also see [recent PR to configure the LaunchAgent for Actions Runner tuning exactly this](https://github.com/actions/runner/pull/2465) - these are cases of dealing with launchd sessions and the services available to them, if I'm reading things right.

This means that a launchd job, say `gitlab-runner`, may not work for your CI because a subset of services are available to that CI run *when running outside of a `gui/$uid` session* (eg: the Dock, appropriate in a GUI session; some security framework services). However, there are use cases that *don't* need the full graphical session. In other words, it depends 😅 - it depends on what you (and your invoked tools) need.

And, yeah, that means we're going to have to work with launchd jobs.

Some relevant links (n.b. in archive docs):

- sessions, domains and how they impact services: 
  https://developer.apple.com/library/archive/technotes/tn2083/_index.html#//apple_ref/doc/uid/DTS10003794-CH1-SUBSECTION10

- general LaunchDaemon, LaunchAgent: 
  https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html

- guidelines for LaunchDaemon vs. LaunchAgent:
  https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/DesigningDaemons.html#//apple_ref/doc/uid/10000172i-SW4-SW9

---

(maybe this works for you @sourcedelica ?)

There are some convenience service helpers in Homebrew for formulas to use. This works for formulas that define a service - and this is [set `gitlab-runner` formula](https://github.com/Homebrew/homebrew-core/blob/73e074779c7cd95287204625606f246a1ea166f8/Formula/gitlab-runner.rb#L39)!

Per https://docs.brew.sh/Manpage#services-subcommand something like this should do the trick:

```sh
#!/usr/bin/env sh
logger -s -t userdata "installing gitlab-runner"
brew install gitlab-runner
logger -s -t userdata "enabling, starting gitlab-runner service"
# re: "sudo" - see https://docs.brew.sh/Manpage#services-subcommand
sudo brew services start gitlab-runner
```

**Comment by sourcedelica** [comment]
I ended up getting the Gitlab runner-installed LaunchAgent working through some hackery.  Basically this was:
1. Set ec2-user as the auto-login user
2. Set the user password and [the auto-login password](https://web.archive.org/web/20171123172641/http://www.brock-family.org/gavin/perl/kcpassword.html) to the same value
3. Reboot
When the machine reboots (and every boot thereafter) the ec2-user is auto-logged in and now the LaunchAgent associated with ec2-user works.  Here are the last few commands in the user-data script:
```
plutil -replace autoLoginUser -string ${user} /Library/Preferences/com.apple.loginwindow.plist
sudo dscl . -passwd /Users/ec2-user "${pw}"
python3 set_kcpassword.py "${pw}"
reboot
```
